File: android/src/main/java/com/nitroplay/hls/HlsProxyRuntimeState.kt
package com.nitroplay.hls

internal class HlsProxyRuntimeState(private val defaultPort: Int = 18181) {
  var port: Int = defaultPort
    private set

  private var isRegistered: Boolean = false
  private var shouldBeRunning: Boolean = false
  private var isExplicitlyStopped: Boolean = false

  @Synchronized
  fun register(): Int {
    isRegistered = true
    shouldBeRunning = true
    isExplicitlyStopped = false
    return port
  }

  @Synchronized
  fun start(nextPort: Int?): Int {
    val resolvedPort = if ((nextPort ?: defaultPort) > 0) {
      nextPort ?: defaultPort
    } else {
      defaultPort
    }

    port = resolvedPort
    isRegistered = true
    shouldBeRunning = true
    isExplicitlyStopped = false
    return port
  }

  @Synchronized
  fun stop() {
    shouldBeRunning = false
    isExplicitlyStopped = true
  }

  @Synchronized
  fun onHostResume(): Boolean {
    return isRegistered && shouldBeRunning && !isExplicitlyStopped
  }

  @Synchronized
  fun onHostDestroy() {
    shouldBeRunning = false
  }

  @Synchronized
  fun shouldEnsureRunningForUse(): Boolean {
    return isRegistered && shouldBeRunning && !isExplicitlyStopped
  }
}

File: ios/hls/HlsProxyRuntimeState.swift
import Foundation

final class HlsProxyRuntimeState {
  private let defaultPort: Int

  private(set) var port: Int
  private var isRegistered = false
  private var shouldBeRunning = false
  private var isExplicitlyStopped = false

  init(defaultPort: Int = 18181) {
    self.defaultPort = defaultPort
    self.port = defaultPort
  }

  @discardableResult
  func register() -> Int {
    isRegistered = true
    shouldBeRunning = true
    isExplicitlyStopped = false
    return port
  }

  @discardableResult
  func start(port requestedPort: Int?) -> Int {
    let resolvedPort = (requestedPort ?? defaultPort) > 0 ? (requestedPort ?? defaultPort) : defaultPort
    port = resolvedPort
    isRegistered = true
    shouldBeRunning = true
    isExplicitlyStopped = false
    return port
  }

  func stop() {
    shouldBeRunning = false
    isExplicitlyStopped = true
  }

  func shouldEnsureRunningForUse() -> Bool {
    isRegistered && shouldBeRunning && !isExplicitlyStopped
  }
}
